opnode/l1/source.go

```diff
@@ -0,0 +1,370 @@
+			if err != nil {
+				return nil, fmt.Errorf("bad block data for block %s: %v", blockRequests[i].Args[0], err)
+			}
+			s.headersCache.Add(info.hash, info)
+			s.transactionsCache.Add(info.hash, txs)
+			allTxLists[requestIndices[i]] = txs
+		}
+	}
+
+	for i := 0; i < len(blockRequests); i++ {
+		if blockRequests[i].Error != nil {
+			return nil, fmt.Errorf("failed to retrieve transactions of block %s in batch of %d blocks: %v", window[i], len(blockRequests), blockRequests[i].Error)
+		}
+	}
+
+	return allTxLists, nil
+}
+
+func (s *Source) L1HeadBlockRef(ctx context.Context) (eth.L1BlockRef, error) {
+	head, err := s.InfoHead(ctx)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch head header: %v", err)
+	}
+	return head.BlockRef(), nil
+}
+
+func (s *Source) L1BlockRefByNumber(ctx context.Context, l1Num uint64) (eth.L1BlockRef, error) {
+	head, err := s.InfoByNumber(ctx, l1Num)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch header by num %d: %v", l1Num, err)
+	}
+	return head.BlockRef(), nil
+}
+
+func (s *Source) L1BlockRefByHash(ctx context.Context, hash common.Hash) (eth.L1BlockRef, error) {
+	block, err := s.InfoByHash(ctx, hash)
+	if err != nil {
+		return eth.L1BlockRef{}, fmt.Errorf("failed to fetch header by hash %v: %w", hash, err)
+	}
+	return block.BlockRef(), nil
+}
+
+// L1Range returns a range of L1 block beginning just after begin, up to max blocks.
+// This batch-requests all blocks by number in the range at once, and then verifies the consistency
+func (s *Source) L1Range(ctx context.Context, begin eth.BlockID, max uint64) ([]eth.BlockID, error) {
+	headerRequests := make([]rpc.BatchElem, max)
+	for i := uint64(0); i < max; i++ {
+		headerRequests[i] = rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{hexutil.EncodeUint64(begin.Number + 1 + i), false},
+			Result: new(*rpcHeader),
+			Error:  nil,
+		}
+	}
+	if err := s.batchCall(ctx, headerRequests); err != nil {
+		return nil, err
+	}
+
+	out := make([]eth.BlockID, 0, max)
+
+	// try to cache everything we have before halting on the results with errors
+	for i := 0; i < len(headerRequests); i++ {
+		result := *headerRequests[i].Result.(**rpcHeader)
+		if headerRequests[i].Error == nil {
+			if result == nil {
+				break // no more headers from here
+			}
+			info, err := result.Info(s.trustRPC)
+			if err != nil {
+				return nil, fmt.Errorf("bad header data for block %s: %v", headerRequests[i].Args[0], err)
+			}
+			s.headersCache.Add(info.hash, info)
+			out = append(out, info.ID())
+			prev := begin
+			if i > 0 {
+				prev = out[i-1]
+			}
+			if prev.Hash != info.parentHash {
+				return nil, fmt.Errorf("inconsistent results from L1 chain range request, block %s not expected parent %s of %s", prev, info.parentHash, info.ID())
+			}
+		} else if errors.Is(headerRequests[i].Error, ethereum.NotFound) {
+			break // no more headers from here
```

Comment:
Does this invariant hold when we don't trust the L1? Couldn't I just return a `null` block somewhere in there and cause the opnode to halt blocks?

Reply:
If the L1 RPC stops responding with blocks, or gives us a null / not-found for whatever reason, it just means that the window-buffer we maintain stops growing. That buffer has to be large enough to slice out the next sequencing window, and extend the L2 chain from that. If the buffer doesn't grow, then we'll indeed halt. But it recovers fine once we see the actual L1 information again (and we reorg if it's a different chain).